Raise error is required attr. is missing [Fix #3455]

reyanshsolis opened the pull request:
Change:
- Raise error is point cloud is missing colors to avoid seg. fault in `RegistrationColoredICP`. [Fix for Issue #3455].

Pending Issues:
- One might still get a segmentation fault if one is using legacy `RegistrationICP` function directly with `estimation = TransformationEstimationType::ColoredICP`. To fix this, some refactoring is required. It will be done in a future PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/intel-isl/open3d/3466)
<!-- Reviewable:end -->

ssheorey reviewed (changes requested):
> Pending Issues:
> 
> * One might still get a segmentation fault if one is using legacy `RegistrationICP` function directly with `estimation = TransformationEstimationType::ColoredICP`. To fix this, some refactoring is required. It will be done in a future PR.

We already have some error checking in that case (Registration.cpp:146):
```
    if ((estimation.GetTransformationEstimationType() ==
                 TransformationEstimationType::PointToPlane ||
         estimation.GetTransformationEstimationType() ==
                 TransformationEstimationType::ColoredICP) &&
        (!target.HasNormals())) {
        utility::LogError(
                "TransformationEstimationPointToPlane and "
                "TransformationEstimationColoredICP "
                "require pre-computed normal vectors for target PointCloud.");
    }
```
Let's add checking for colors as well.

reyanshsolis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3466)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @ssheorey)

___
*[cpp/open3d/pipelines/registration/ColoredICP.cpp, line 126 at r2](https://reviewable.io/reviews/intel-isl/open3d/3466#-Ma0AfPYozkPq-wxwGI0-r2-126:-Ma4tc3L9UZ2SVgA2lZs:b2nhky0) ([raw file](https://github.com/intel-isl/open3d/blob/0acefc2a849df1bacaaf669f61a9e1a8b3242c5e/cpp/open3d/pipelines/registration/ColoredICP.cpp#L126)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Does the source also need normals?
</blockquote></details>

No.
___
*[cpp/open3d/pipelines/registration/ColoredICP.cpp, line 244 at r2](https://reviewable.io/reviews/intel-isl/open3d/3466#-Ma0AfPYozkPq-wxwGI0-r2-244:-Ma4teBm04mfaKBdc6OZ:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/0acefc2a849df1bacaaf669f61a9e1a8b3242c5e/cpp/open3d/pipelines/registration/ColoredICP.cpp#L244)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

/* = TransformationEstimationForColoredICP() */
</blockquote></details>

Done.
___
*[cpp/open3d/pipelines/registration/ColoredICP.cpp, line 247 at r2](https://reviewable.io/reviews/intel-isl/open3d/3466#-Ma0AfPYozkPq-wxwGI0-r2-247:-Ma4teYX8Qc9eVUOznLR:b2nhky0) ([raw file](https://github.com/intel-isl/open3d/blob/0acefc2a849df1bacaaf669f61a9e1a8b3242c5e/cpp/open3d/pipelines/registration/ColoredICP.cpp#L247)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Does the source also need normals?
</blockquote></details>

No.


<!-- Sent from Reviewable.io -->

reyanshsolis commented:
> > Pending Issues:
> > 
> > * One might still get a segmentation fault if one is using legacy `RegistrationICP` function directly with `estimation = TransformationEstimationType::ColoredICP`. To fix this, some refactoring is required. It will be done in a future PR.
> 
> We already have some error checking in that case (Registration.cpp:146):
> 
> ```
>     if ((estimation.GetTransformationEstimationType() ==
>                  TransformationEstimationType::PointToPlane ||
>          estimation.GetTransformationEstimationType() ==
>                  TransformationEstimationType::ColoredICP) &&
>         (!target.HasNormals())) {
>         utility::LogError(
>                 "TransformationEstimationPointToPlane and "
>                 "TransformationEstimationColoredICP "
>                 "require pre-computed normal vectors for target PointCloud.");
>     }
> ```
> 
> Let's add checking for colors as well.

Only a colored property check won't prevent the user from running into an error (seg fault), as `colored ICP` also requires a pre-computed colored gradient. Therefore if one directly uses `RegistrationICP` with `TransformationEstimationType::ColoredICP`, and if one is using `RegistrationColoredICP`, then checks are performed there. 

Solution:
1. For colored ICP, use a local static copy of `RegistrationICP` function (but in that case either `GetRegistrationResultAndCorrepondences` function needs to be exposed or copied to the `ColoredICP` file). And, throw error is someone calls the `RegistrationICP` function with `TransformationEstimationType::ColoredICP` directly.

2. Depreciate `RegistrationColoredICP.cpp and .h files`, move the relevant functions to `Registration` and `TransformationEstimation`, and make `RegistrationICP` a common function (users can just run it with `TransformationEstimationType::ColoredICP` to use `ColoredICP` like they currently do with `PointToPoint` and `PointToPlane`).
[I would personally prefer this one]. 

I plan to do this refactoring in the near future if the team concurs.

ssheorey commented:
Option 2 sounds good to me. LGTM for now.

reyanshsolis commented on cpp/open3d/pipelines/registration/ColoredICP.cpp:
```diff
@@ -231,6 +244,19 @@ RegistrationResult RegistrationColoredICP(
         /*TransformationEstimationForColoredICP()*/,
```
Done.

reyanshsolis commented on cpp/open3d/pipelines/registration/ColoredICP.cpp:
```diff
@@ -118,9 +118,22 @@ Eigen::Matrix4d TransformationEstimationForColoredICP::ComputeTransformation(
+                "No correspondences found between source and target "
+                "pointcloud.");
+    }
+    if (!target.HasNormals()) {
```
No, not required.

reyanshsolis commented on cpp/open3d/pipelines/registration/ColoredICP.cpp:
```diff
@@ -231,6 +244,19 @@ RegistrationResult RegistrationColoredICP(
         /*TransformationEstimationForColoredICP()*/,
         const ICPConvergenceCriteria
                 &criteria /* = ICPConvergenceCriteria()*/) {
+    if (!target.HasNormals()) {
```
No, not required.

reyanshsolis commented:
> Option 2 sounds good to me. LGTM for now (except the earlier question about source normals).

Thanks for your review and approval on option 2, I will start working on it. 

I had addressed your comment on Reviewable. We don't require source normals. I also verified this by searching if `source.normals_` was being used anywhere in the code.
